Baseline Modeling - Car Price Prediction

In [9]:
# Step 1 – Setup

# Import core libraries
import pandas as pd
import numpy as np

# Model & preprocessing tools
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Regression models
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

# Load the cleaned dataset
df = pd.read_csv('E:/CY Tech/Big Data project/Project 1 Car Price Prediction Multiple Linear Regression/Data/Processed/car_price_clean.csv')

# Quick shape and preview
print("Dataset shape:", df.shape)
display(df.head())

Dataset shape: (205, 27)


,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price,brand
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0,alfa-romero
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0,alfa-romero
2,3,1,alfa-romero quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0,alfa-romero
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0,audi
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0,audi


1- Train/Test Split Strategy:

- Use 80% of the data for training and 20% for testing.
- Fix random_state for reproducibility.

In [2]:
# Separate target and features
y = df['price']
X = df.drop('price', axis=1)

# Identify numeric and categorical columns
numeric_cols = X.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X.select_dtypes(include=['object']).columns.tolist()

print("Numeric columns:", numeric_cols)
print("Categorical columns:", categorical_cols)

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

# Preprocessing: scale numerics + one-hot encode categoricals
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Train-test split (80/20 split)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Training set size:", X_train.shape)
print("Test set size:", X_test.shape)

# Fit on training data, transform both train and test
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

print("Processed X_train shape:", X_train_processed.shape)
print("Processed X_test shape:", X_test_processed.shape)

Numeric columns: ['car_ID', 'symboling', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg']
Categorical columns: ['CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'enginetype', 'cylindernumber', 'fuelsystem', 'brand']
Training set size: (164, 26)
Test set size: (41, 26)
Processed X_train shape: (164, 198)
Processed X_test shape: (41, 198)


2- Data Preprocessing:

- Apply one-hot encoding for categorical features.
- Apply StandardScaler for numerical features.
- Ensure both transformations are applied consistently to train and test sets.

In [3]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

# Define which columns are numeric and which are categorical
numeric_cols = X.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X.select_dtypes(include=['object']).columns.tolist()

print("Numeric columns:", numeric_cols)
print("Categorical columns:", categorical_cols)

# ColumnTransformer: scale numeric, one-hot encode categorical
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numeric_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

# Fit/transform training data and transform test data
X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

print("Processed X_train shape:", X_train_processed.shape)
print("Processed X_test shape:", X_test_processed.shape)

Numeric columns: ['car_ID', 'symboling', 'wheelbase', 'carlength', 'carwidth', 'carheight', 'curbweight', 'enginesize', 'boreratio', 'stroke', 'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg']
Categorical columns: ['CarName', 'fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'enginetype', 'cylindernumber', 'fuelsystem', 'brand']
Processed X_train shape: (164, 198)
Processed X_test shape: (41, 198)


3- Baseline Regression Models (Without Optimization):

Models to test:

- Linear Regression
- Ridge Regression
- Lasso Regression
- Decision Tree Regressor
- Random Forest Regressor

In [4]:
# Define all baseline models
models = {
    "Linear Regression": LinearRegression(),
    "Ridge Regression": Ridge(),
    "Lasso Regression": Lasso(),
    "Decision Tree": DecisionTreeRegressor(random_state=42),
    "Random Forest": RandomForestRegressor(random_state=42)
}

# Dictionary to store predictions
predictions = {}

# Train and predict with each model
for name, model in models.items():
    print(f"Training: {name}")
    model.fit(X_train_processed, y_train)
    y_pred = model.predict(X_test_processed)
    predictions[name] = y_pred
    print(f"  Done.")

print("\n All baseline models trained successfully.")

Training: Linear Regression
  Done.
Training: Ridge Regression
  Done.
Training: Lasso Regression
  Done.
Training: Decision Tree
  Done.
Training: Random Forest


C:\Users\fdame\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:656: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 2787384.9491672963, tolerance: 978060.5331304879
  model = cd_fast.sparse_enet_coordinate_descent(


  Done.

 All baseline models trained successfully.


4- Model Evaluation Metrics:

For each model, calculate and compare:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

In [5]:
results = []

for name, y_pred in predictions.items():
    mae = mean_absolute_error(y_test, y_pred)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred))
    r2 = r2_score(y_test, y_pred)
    
    results.append({
        "Model": name,
        "MAE": round(mae, 2),
        "RMSE": round(rmse, 2),
        "R²": round(r2, 4)
    })

# Convert results to DataFrame
results_df = pd.DataFrame(results)

results_df

,Model,MAE,RMSE,R²
0,Linear Regression,3152.93,4961.85,0.6881
1,Ridge Regression,1823.77,2762.44,0.9033
2,Lasso Regression,1913.83,2845.79,0.8974
3,Decision Tree,2098.49,3427.12,0.8512
4,Random Forest,1380.47,1938.93,0.9524


5- Results Comparison Table:

- Create a simple table summarizing the performance of all baseline models.

In [6]:
# Results Comparison Table

# Sort by R² score (highest is better)
results_sorted = results_df.sort_values(by='R²', ascending=False).reset_index(drop=True)

print(" Baseline Model Performance (sorted by R²):")
results_sorted

 Baseline Model Performance (sorted by R²):


,Model,MAE,RMSE,R²
0,Random Forest,1380.47,1938.93,0.9524
1,Ridge Regression,1823.77,2762.44,0.9033
2,Lasso Regression,1913.83,2845.79,0.8974
3,Decision Tree,2098.49,3427.12,0.8512
4,Linear Regression,3152.93,4961.85,0.6881


In [7]:
results_df.to_csv('E:/CY Tech/Big Data project/Project 1 Car Price Prediction Multiple Linear Regression/Data/Processed/baseline_results.csv', index=False)

*Baseline Conclusion*

- The best baseline model was the Random Forest Regressor.
- It achieved an R² of 0.9524, RMSE of 1938.93, and MAE of 1380.47.
- It performed better than the linear models because it captures non-linear relationships and handles interactions between features more effectively than simple regression methods.